romtools/workflows/residual_evaluator.py

```diff
@@ -0,0 +1,184 @@
+
+
+class SteadyResidualEvaluator(Protocol):
+    """
+    Baseline steady residual evaluator protocol
+    """
+
+    def load_projected_full_solution(self, filename: str) -> np.ndarray:
+        """
+        Reads the full-order model solution from the specified filename
+
+        Args:
+            filename (str): filename of the file containing the full-order model solution data
+
+        Returns:
+            `np.ndarray`: The projected full-order solution in a 1-dimensional array
+
+        """
+        pass
+
+    def evaluate_full_residual(self, run_directory: str, full_model_directory: str, reduced_state: np.ndarray) -> np.ndarray:
+        """
+        Evaluate the full-order model residual corresponding to a full state reconstructed from
+        a reduced state reduced_state
+
+        Args:
+            run_directory (str): Absolute path to directory in which residual is being computed.
+            full_model_directory (str): Absolute path to directory in which the full model data was computed.
+            reduced_state (np.ndarray): 1-dimensional reduced state vector.
+
+        Returns:
+            `np.ndarray`: The full-order residual in tensor form.
+        """
+        # TODO should this just return a list of directories instead?
+        pass
+
+
+# TODO populate_run_directory method as well?
+class UnsteadyResidualEvaluator(Protocol):
+    """
+    Baseline unsteady residual evaluator protocol
+    """
+
+    def load_projected_full_solutions(self, filename: str) -> (np.ndarray, np.ndarray):
+        """
+        Reads the full-order model solutions and time steps from the specified filename
+
+        Args:
+            filename (str): filename of the file containing the full-order model solution data
+
+        Returns:
+            `np.ndarray`: The projected full-order solution in a 2-dimensional array
+            `np.ndarray`: The corresponding solution time stamps in a 1-dimensional array
+
+        """
+        pass
+
+    def evaluate_full_residuals(self, run_directory: str, full_model_directory: str, reduced_states: np.ndarray, times: np.ndarray) -> np.ndarray:
+        """
+        Evaluate the full-order model residuals corresponding to full states reconstructed from
+        an array of reduced states, reduced_states
+
+        Args:
+            run_directory (str): Absolute path to directory in which residual is being computed.
+            full_model_directory (str): Absolute path to directory in which the full model data was computed.
+            reduced_state (np.ndarray): 2-dimensional reduced state vector.
+            times (np.ndarray): 1-dimensional vector of time stamps.
+
+        Returns:
+            `np.ndarray`: The full-order residual in tensor form, should be 3-dimensional, even for a single time step
+        """
+        # TODO should this just return a list of directories instead?
+        pass
+
+
+# TODO populate_run_directory method as well?
```

Comment:
I think we're OK without a `populate_run_directory` method because I could envision a scenario where you just handle the residuals in memory and don't write it out to a file.

Reply:
Great, thanks!